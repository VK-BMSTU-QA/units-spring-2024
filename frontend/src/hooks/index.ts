export * from './useCurrentTime';
export * from './useProducts';
