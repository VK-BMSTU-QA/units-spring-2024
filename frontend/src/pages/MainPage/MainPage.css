.main-page {
    width: 100vw;
    height: 100vh;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
}

.main-page__title {
    padding: 20px 10px;
}

.main-page__parameters {
    max-width: 70vw;
    display: flex;
    justify-content: space-between;
    align-items: center;

    gap: 10vw;
}

