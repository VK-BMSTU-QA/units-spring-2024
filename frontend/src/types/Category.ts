export type Category = 'Электроника' | 'Для дома' | 'Одежда';
