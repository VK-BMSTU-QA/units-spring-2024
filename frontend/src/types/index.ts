export * from './Symbol';
export * from './Product';
export * from './Category';
