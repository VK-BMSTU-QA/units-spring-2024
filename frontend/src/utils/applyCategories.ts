import type { Category, Product } from '../types';

export const applyCategories = (
    products: Product[],
    categories: Category[]
): Product[] => {
    if (!categories.length) return products;

    return products.filter((product) => categories.includes(product.category));
};
