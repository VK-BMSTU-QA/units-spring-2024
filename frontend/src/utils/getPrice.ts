import type { PriceSymbol } from '../types';

export const getPrice = (value: number, symbol: PriceSymbol = 'â‚½'): string =>
    `${value.toLocaleString('ru-RU')} ${symbol}`;

    