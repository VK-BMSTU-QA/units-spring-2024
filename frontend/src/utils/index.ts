export * from './updateCategories';
export * from './getPrice';
export * from './applyCategories';

