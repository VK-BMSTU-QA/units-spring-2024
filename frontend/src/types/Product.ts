import { Category } from './Category';
import { PriceSymbol } from './Symbol';

export type Product = {
    id: number;
    name: string;
    description: string;
    price: number;
    priceSymbol?: PriceSymbol;
    imgUrl?: string;
    category: Category;
};
