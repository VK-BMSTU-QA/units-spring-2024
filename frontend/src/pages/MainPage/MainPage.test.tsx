import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MainPage } from './MainPage';
import { useCurrentTime, useProducts } from '../../hooks';
import { applyCategories, updateCategories } from '../../utils';
import { Product, Category } from '../../types';


const products: Product[] = [
    {
        id: 1,
        name: 'IPhone 14 Pro',
        description: 'Latest iphone, buy it now',
        price: 999,
        priceSymbol: '$',
        category: 'Электроника',
        imgUrl: '/iphone.png',
    },
    {
        id: 2,
        name: 'Костюм гуся',
        description: 'Запускаем гуся, работяги',
        price: 1000,
        priceSymbol: '₽',
        category: 'Одежда',
    },
    {
        id: 3,
        name: 'Настольная лампа',
        description: 'Говорят, что ее использовали в pixar',
        price: 699,
        category: 'Для дома',
        imgUrl: '/lamp.png',
    },
    {
        id: 4,
        name: 'Принтер',
        description: 'Незаменимая вещь для студента',
        price: 7000,
        category: 'Электроника',
    },
]

jest.mock('../../hooks', () => ({
    useCurrentTime: jest.fn(() => '00:00:00'),
    useProducts: jest.fn(() => products)
}));

jest.mock('../../utils', () => ({
    applyCategories: jest.fn(() => [
        {
          id: 1,
          name: 'Телефон',
          description: 'Описание продукта 1',
          price: 100,
          category: 'Электроника'
        },
        {
          id: 4,
          name: 'Ноутбук',
          description: 'Описание продукта 4',
          price: 100,
          category: 'Электроника'
        }
      ]),
    getPrice:  jest.fn(),
    updateCategories: jest.fn(() => ['Одежда']),
}));

afterEach(jest.clearAllMocks);
describe('MainPage test', () => {
    it('should render correctly', () => {
        const rendered = render(<MainPage />);

        expect(rendered.asFragment()).toMatchSnapshot();
    });


    it('should called categories butten after click', () => {
        const rendered = render(<MainPage />);
        expect(updateCategories).toHaveBeenCalledTimes(0);

        const categoryButton = rendered.getByText('Одежда', {
            selector: '.categories__badge',
        });
        fireEvent.click(categoryButton);
        expect(updateCategories).toHaveBeenCalledTimes(1);
    });

    it('should called when onload page', () => {
        render(<MainPage />);
        expect(useProducts).toHaveBeenCalledTimes(1);
        expect(useCurrentTime).toHaveBeenCalledTimes(1);
        expect(applyCategories).toHaveBeenCalledWith(products, []);
    });

});