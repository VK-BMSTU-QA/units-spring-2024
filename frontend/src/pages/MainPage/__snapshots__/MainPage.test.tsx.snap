// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage test should render correctly 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3>
      00:00:00
    </h3>
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Телефон
        </h2>
        <div
          class="product-card__description"
        >
          Описание продукта 1
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Ноутбук
        </h2>
        <div
          class="product-card__description"
        >
          Описание продукта 4
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
